nexus/src/app/vpc.rs

```diff
@@ -315,18 +358,366 @@ impl super::Nexus {
+        // subnets for things that are specified as host filters as well.
+        let mut instances: HashSet<Name> = HashSet::new();
+        let mut subnets: HashSet<Name> = HashSet::new();
+        let mut vpcs: HashSet<Name> = HashSet::new();
+        for rule in rules {
+            for target in &rule.targets {
+                match &target.0 {
+                    external::VpcFirewallRuleTarget::Instance(name) => {
+                        instances.insert(name.clone().into());
+                    }
+                    external::VpcFirewallRuleTarget::Subnet(name) => {
+                        subnets.insert(name.clone().into());
+                    }
+                    external::VpcFirewallRuleTarget::Vpc(name) => {
+                        if name != vpc.name() {
+                            return Err(Error::invalid_request(
+                                "cross-VPC firewall target unsupported",
+                            ));
+                        }
+                        vpcs.insert(name.clone().into());
+                    }
+                    // We don't need to resolve anything for Ip(Net)s.
+                    external::VpcFirewallRuleTarget::Ip(_) => (),
+                    external::VpcFirewallRuleTarget::IpNet(_) => (),
+                }
+            }
+
+            for host in rule.filter_hosts.iter().flatten() {
+                match &host.0 {
+                    external::VpcFirewallRuleHostFilter::Instance(name) => {
+                        instances.insert(name.clone().into());
+                    }
+                    external::VpcFirewallRuleHostFilter::Subnet(name) => {
+                        subnets.insert(name.clone().into());
+                    }
+                    external::VpcFirewallRuleHostFilter::Vpc(name) => {
+                        if name != vpc.name() {
+                            return Err(Error::invalid_request(
+                                "cross-VPC firewall host filter unsupported",
+                            ));
+                        }
+                        vpcs.insert(name.clone().into());
+                    }
+                    // We don't need to resolve anything for Ip(Net)s.
+                    external::VpcFirewallRuleHostFilter::Ip(_) => (),
+                    external::VpcFirewallRuleHostFilter::IpNet(_) => (),
+                }
+            }
+        }
+
+        // Resolve named instances, VPCs, and subnets.
+        // TODO-correctness: It's possible the resolving queries produce
+        // inconsistent results due to concurrent changes. They should be
+        // transactional.
+        type NetMap = HashMap<external::Name, Vec<IpNetwork>>;
+        type NicMap = HashMap<external::Name, Vec<NetworkInterface>>;
+        let no_networks: Vec<IpNetwork> = Vec::new();
+        let no_interfaces: Vec<NetworkInterface> = Vec::new();
+
+        let mut instance_interfaces: NicMap = HashMap::new();
```

Comment:
Excellent question; 1fa97ab8 does explicit de-duplication on the NICs. (If `sled_agent_client::types::NetworkInterface` were `Hash`, this could be simplified by accumulating directly into a `HashSet`; please LMK if you know how to do that.)

Reply:
I think you can implement `Hash` manually, probably by hashing all the fields. Not sure if it's worth it, up to you at this point, since I don't want to further delay this PR unnecessarily.